### PR TITLE
🦑  ☔️  Refactor weather weather for open weather api 👿

### PR DESCRIPTION
**Pull request by lisamittmann**
- add new open weather api service plus tests
- adjust weather service to include new open weather api service + tests
- remove unused code from meta weather api
- adjust terraform and pipeline to reflect new env var

### CODE REVIEW

File: .github/workflows/deployment.yml [hunk 1]
```diff
@@ -37,9 +37,11 @@ jobs:
       - name: Terraform Plan
         env:
           MONGODB_URI: ${{ secrets.MONGODB_URI }}
+          OPENWEATHER_API_KEY: ${{ secrets.OPENWEATHER_API_KEY }}
         run: |
           cd iac/terraform
           printf "mongodburi = \"$MONGODB_URI\"" > terraform.tfvars
+          printf "openweatherkey = \"OPENWEATHER_API_KEY\"" >> terraform.tfvars
```
Comment (jamarob) [inline .github/workflows/deployment.yml]: I think `echo` would be a better choice than `printf` as it adds a newline?

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/config/OpenWeatherKeyConfig.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.context.annotation.Configuration;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Configuration
+@ConfigurationProperties(prefix = "openweather.key")
+public class OpenWeatherKeyConfig {
+
+    private String apiKey;
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/config/OpenWeatherKeyConfig.java]: Resulting key name is `openweather.key.apiKey` why not simply `openweather.key`?

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherContent.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+@AllArgsConstructor
+@NoArgsConstructor
+public class WeatherContent {
+    @JsonProperty("main")
+    private String weatherClass;
+    @JsonProperty("description")
+    private String weatherDescription;
+    @JsonProperty("icon")
+    private String iconId;
+}
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherContent.java]: `Content` is way to generic to help me understand what this data represents 🤔 I would not prefix the members with `weather` - I see the need for `weatherClass` as `class` is a Java keyword, but now I wonder what the class of weather is?

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherTemperature.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class WeatherTemperature {
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherTemperature.java]: Why not simply `Temperature`?

`Temperatue` post fix doesn't add that much, imho.

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/service/OpenWeatherApiService.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.client.RestTemplate;
+
+import java.util.Optional;
+
+@Service
+public class OpenWeatherApiService {
+
+    private final RestTemplate restTemplate;
+    private final OpenWeatherKeyConfig openWeatherKeyConfig;
+    private String baseUrl = "http://api.openweathermap.org/data/2.5/weather";
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/service/OpenWeatherApiService.java]:
```suggestion
    private static final String baseUrl = "http://api.openweathermap.org/data/2.5/weather";
```

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java [hunk 6]
```diff
@@ -37,35 +32,24 @@ public ProvinceCapitalWeatherData getProvinceCapitalWeatherData(String province)
...
-
-        if(weatherDataForProvinceCapital.isEmpty()) {throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Weather API not available");}
-
-        return getProvinceCapitalWeatherDataFromWeatherData(provinceData.get().getCapital(), weatherDataForProvinceCapital.get());
-    }
-
-    public Optional<WeatherData> getWeatherDataForProvinceCapital(String capitalWoeid) {
-        List<WeatherData> weatherData = metaWeatherApiService.getWeatherData(capitalWoeid);
+        String capital = provinceData.get().getCapital();
+        Optional<Weather> weatherData = openWeatherApiService.getWeatherForProvinceCapital(capital);
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java]:
```suggestion
        Weather weather = openWeatherApiService.getWeatherForProvinceCapital(capital)
            .orThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Weather API not available"));
```

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java [hunk 7]
```diff
@@ -37,35 +32,24 @@ public ProvinceCapitalWeatherData getProvinceCapitalWeatherData(String province)
...
 
-        String currentDay = currentDateUtils.getCurrentDay();
+        if(weatherData.isEmpty()) {throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Weather API not available");}
 
-        return weatherData.stream().filter(dataSet -> dataSet.getApplicableDate().equals(currentDay)).findAny();
+        return getProvinceCapitalWeatherDataFromWeatherData(provinceData.get().getCapital(), weatherData.get());
     }
 
-    public ProvinceCapitalWeatherData getProvinceCapitalWeatherDataFromWeatherData(String capital, WeatherData data) {
+    public ProvinceCapitalWeatherData getProvinceCapitalWeatherDataFromWeatherData(String capital, Weather data) {
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java]:
```suggestion
    public ProvinceCapitalWeatherData getProvinceCapitalWeatherDataFromWeatherData(String capital, Weather weatherData) {
```
`data` is way too generic.

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java [hunk 8]
```diff
@@ -37,35 +32,24 @@ public ProvinceCapitalWeatherData getProvinceCapitalWeatherData(String province)
...
                 capital,
-                WeatherCategories.getWeatherCategory(data.getWeatherStateAbbreviation().toUpperCase()).get(),
-                (int)Math.round(data.getMinTemperature()),
-                (int)Math.round(data.getMaxTemperature()),
-                (int)Math.round(data.getCurrentTemperature()),
-                "https://www.metaweather.com/static/img/weather/" + data.getWeatherStateAbbreviation() + ".svg"
+                data.getWeather().get(0).getWeatherDescription(),
+                (int)Math.round(data.getTemperature().getMinTemperature()),
+                (int)Math.round(data.getTemperature().getMaxTemperature()),
+                (int)Math.round(data.getTemperature().getCurrentTemperature()),
```
Comment (jamarob) [inline backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java]: Why not return a rounded double and not cast here? Also using a builder safes me from thinking what each value means.

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java [hunk 9]
```diff
@@ -58,10 +57,9 @@ public void setup() {
     public void getProvinceCapitalWeatherDataShouldReturnWeatherData() {
         // Given
         String province = "Hamburg";
-        provinceMongoDb.save(new ProvinceData("Hamburg", "Hamburg", "656958"));
-        when(currentDateUtils.getCurrentDay()).thenReturn("2021-03-21");
-        when(restTemplate.getForEntity("https://www.metaweather.com/api/location/656958", WeatherDataWrapper.class))
-                .thenReturn(ResponseEntity.ok(new WeatherDataWrapper(getWeatherDataList())));
+        provinceMongoDb.save(new ProvinceData("Hamburg", "Hamburg"));
+        when(restTemplate.getForEntity("http://api.openweathermap.org/data/2.5/weather?q=Hamburg&appid=awesomeApiKey&units=metric&lang=DE", Weather.class))
```
Comment (jamarob) [inline backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java]: maybe extract the url to a variable for better readability?

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java [hunk 10]
```diff
@@ -70,11 +68,11 @@ public void getProvinceCapitalWeatherDataShouldReturnWeatherData() {
         assertThat(response.getStatusCode(), is(HttpStatus.OK));
         assertThat(response.getBody(), is(new ProvinceCapitalWeatherData(
```
Comment (jamarob) [inline backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java:69]: Using a builder here helps me understand what's going on. So this seems to be a DTO maybe rename to `ProvinceCapitalWeatherDto` ?

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java [hunk 11]
```diff
@@ -97,9 +95,8 @@ public void getProvinceCapitalWeatherDataShouldThrowErrorForUnknownProvince(){
     public void getProvinceCapitalWeatherdataShouldThrowErrorWhenApiUnavailable() {
         // Given
         String province = "Hamburg";
-        provinceMongoDb.save(new ProvinceData("Hamburg", "Hamburg", "656958"));
-        when(currentDateUtils.getCurrentDay()).thenReturn("2021-03-21");
-        when(restTemplate.getForEntity("https://www.metaweather.com/api/location/656958", WeatherDataWrapper.class))
+        provinceMongoDb.save(new ProvinceData("Hamburg", "Hamburg"));
+        when(restTemplate.getForEntity("http://api.openweathermap.org/data/2.5/weather?q=Hamburg&appid=awesomeApiKey&units=metric&lang=DE", Weather.class))
```
Comment (jamarob) [inline backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java]: See above

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java [hunk 12]
```diff
@@ -110,42 +107,9 @@ public void getProvinceCapitalWeatherdataShouldThrowErrorWhenApiUnavailable() {
...
-                        .applicableDate("2021-03-23")
-                        .minTemperature(3.875)
-                        .maxTemperature(13.86)
-                        .currentTemperature(12.065000000000001)
-                        .windSpeed(4.708858778127734)
-                        .humidity(71)
-                        .build()
-        );
+    private Weather getWeather() {
+        return new Weather(List.of(new WeatherContent("Rain", "Leichter Regen", "01d")),
```
Comment (jamarob) [inline backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java]: Builder ;) Don't force me to browse your models to understand what `"01d"` is supposed to be.

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/openweatherapi/service/OpenWeatherApiServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class OpenWeatherApiServiceTest {
+
+    private final RestTemplate restTemplate = mock(RestTemplate.class);
+    private final OpenWeatherKeyConfig openWeatherKeyConfig = mock(OpenWeatherKeyConfig.class);
+    private final OpenWeatherApiService openWeatherApiService = new OpenWeatherApiService(restTemplate, openWeatherKeyConfig);
+    private String baseUrl = "http://api.openweathermap.org/data/2.5/weather";
```
Comment (jamarob) [inline backend/src/test/java/de/neuefische/teddyscoronadiaries/openweatherapi/service/OpenWeatherApiServiceTest.java]:
```suggestion
    private static final String baseUrl = "http://api.openweathermap.org/data/2.5/weather";
```

**Review by jamarob — APPROVED** [review]
Gj! 🥇 Code looks fine! 

Imho you could improve some names. 

I know you're not a big fan of the builder pattern but it really helps with readability 😉 

Can't comment on the terraform files as I missed your workshop 🙏

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherContent.java (hunk 3 above)
Comment (lisamittmann) [reply backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherContent.java]: the best class of weather is of course `Sunshine ☀️`

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java (hunk 12 above)
Comment (lisamittmann) [reply backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java]: fine, I'll use a builder just for you just so you know 🙈

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java (hunk 8 above)
Comment (lisamittmann) [reply backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java]: don't wanna have to round / trim / whatever in the frontend so I do it here ... also I do not want to use a builder here to ensure that all values are defined for the frontend 😝

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherContent.java (hunk 3 above)
Comment (jamarob) [reply backend/src/main/java/de/neuefische/teddyscoronadiaries/openweatherapi/model/WeatherContent.java]: Okay 😆

File: backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java (hunk 8 above)
Comment (jamarob) [reply backend/src/main/java/de/neuefische/teddyscoronadiaries/service/WeatherService.java]: Yeah, but you have a test for that ;)

File: backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java (hunk 12 above)
Comment (jamarob) [reply backend/src/test/java/de/neuefische/teddyscoronadiaries/controller/WeatherControllerTest.java]: 🙏

**Pull request merged**